ledger/participant-integration-api/src/main/scala/platform/indexer/ExecuteUpdate.scala

```diff
@@ -0,0 +1,416 @@
+          "updateSourceDescription" -> sourceDescription.getOrElse(""),
+        )
+      case PublicPackageUploadRejected(submissionId, _, rejectionReason) =>
+        Map(
+          "updateSubmissionId" -> submissionId,
+          "updateRejectionReason" -> rejectionReason,
+        )
+      case TransactionAccepted(optSubmitterInfo, transactionMeta, _, transactionId, _, _, _) =>
+        Map(
+          "updateTransactionId" -> transactionId,
+          "updateLedgerTime" -> transactionMeta.ledgerEffectiveTime.toInstant.toString,
+          "updateWorkflowId" -> transactionMeta.workflowId.getOrElse(""),
+          "updateSubmissionTime" -> transactionMeta.submissionTime.toInstant.toString,
+        ) ++ optSubmitterInfo
+          .map(info =>
+            Map(
+              "updateSubmitter" -> loggingContextPartiesValue(info.actAs),
+              "updateApplicationId" -> info.applicationId,
+              "updateCommandId" -> info.commandId,
+              "updateDeduplicateUntil" -> info.deduplicateUntil.toString,
+            )
+          )
+          .getOrElse(Map.empty)
+      case CommandRejected(_, submitterInfo, reason) =>
+        Map(
+          "updateSubmitter" -> loggingContextPartiesValue(submitterInfo.actAs),
+          "updateApplicationId" -> submitterInfo.applicationId,
+          "updateCommandId" -> submitterInfo.commandId,
+          "updateDeduplicateUntil" -> submitterInfo.deduplicateUntil.toString,
+          "updateRejectionReason" -> reason.description,
+        )
+    }
+
+  private def loggingContextPartiesValue(parties: List[Party]): String =
+    parties.mkString("[", ", ", "]")
+}
+
+class PipelinedExecuteUpdate(
+    private[indexer] val ledgerDao: LedgerDao,
+    private[indexer] val metrics: Metrics,
+    private[indexer] val participantId: v1.ParticipantId,
+)(implicit val executionContext: ExecutionContext, val loggingContext: LoggingContext)
+    extends ExecuteUpdate {
+
+  private val insertTransactionState: PreparedUpdate => Future[PreparedUpdate] = {
+    case offsetUpdate @ PreparedTransactionInsert(_, _, preparedInsert) =>
+      Timed.future(
+        metrics.daml.index.db.storeTransactionState,
+        ledgerDao.storeTransactionState(preparedInsert).map(_ => offsetUpdate),
+      )
+    case offsetUpdate => Future.successful(offsetUpdate)
+  }
+
+  private val insertTransactionEvents: PreparedUpdate => Future[PreparedUpdate] = {
+    case offsetUpdate @ PreparedTransactionInsert(_, _, preparedInsert) =>
+      Timed.future(
+        metrics.daml.index.db.storeTransactionEvents,
+        ledgerDao
+          .storeTransactionEvents(preparedInsert)
+          .map(_ => offsetUpdate),
+      )
+    case offsetUpdate => Future.successful(offsetUpdate)
+  }
+
+  private[indexer] val flow: ExecuteUpdateFlow =
+    Flow[OffsetUpdate]
+      .mapAsync(1)(prepareUpdate)
+      .mapAsync(1)(insertTransactionState)
+      .mapAsync(1)(insertTransactionEvents)
```

Comment:
Indeed, this is a way to go forward.

Reply:
We could batch through a [.conflate()](https://doc.akka.io/docs/akka/current/stream/operators/Source-or-Flow/conflate.html) stream operator. 

But I think this would be counter-productive. The point of this PR is to allow all three updates to run asynchronously, so that the next transaction can start writing contracts as soon as the first transaction is done with contracts and starts writing events.